### PR TITLE
Variable Cost Refactor Part 1: Function Data

### PR DESCRIPTION
**Pull request by GabrielKS**
The first part of a significant refactor of Sienna cost data structures. See https://github.com/NREL-Sienna/PowerSystems.jl/pull/1056 for more details. This PR consists of a new type `FunctionData`, a new set of concrete subtype structs to represent various functions, and the necessary refactoring to use them rather than `IS.PWL` and such in other InfrastructureSystems data structures. I also performed some minor code deduplication.

IS, PSY, and PSB tests pass when using the appropriate branches of each plus PowerSystemsTestData:
Screenshot of passing tests

### CODE REVIEW

File: src/utils/utils.jl [hunk 1]
```diff
@@ -464,8 +496,10 @@ function transform_array_for_hdf(data::Vector{POLYNOMIAL})
     return t_lin_cost
 end
 
-function transform_array_for_hdf(data::SortedDict{Dates.DateTime, Vector{PWL}})
-    quad_cost = hcat(values(data)...)
+function transform_array_for_hdf(
+    data::SortedDict{Dates.DateTime, Vector{PiecewiseLinearPointData}},
+)
+    quad_cost = hcat([get_points.(v) for v in values(data)]...)
```
Comment (jd-lara) [inline src/utils/utils.jl:502]: Why is this called `quad_cost` for a PWL data? I see that the name is prior to this PR but it seems incorrect.

File: src/utils/utils.jl [hunk 2]
```diff
@@ -481,7 +515,10 @@ function transform_array_for_hdf(data::SortedDict{Dates.DateTime, Vector{PWL}})
     return t_quad_cost
 end
 
-function transform_array_for_hdf(data::Vector{PWL})
+# TODO: old code here does not properly handle data with different numbers of points
```
Comment (jd-lara) [inline src/utils/utils.jl:518]: This is ok. It was meant like that for proper storage in the HDF5 file

**Review by jd-lara — APPROVED** [review]
This PR looks ok to me, there is some pending work for the MarketBid Cost that requires scoping but we already decided to refactor that at a later stage.

File: src/utils/utils.jl (hunk 1 above)
Comment (GabrielKS) [reply src/utils/utils.jl:502]: It does seem incorrect. I'll change it

**Pull request merged**